src/CommandingOfficers/Commander.java

```diff
@@ -96,11 +97,14 @@ protected void addCommanderAbility(CommanderAbility ca)
    */
   public void changeCombatContext(CombatContext instance)
   {}
-  /**
-   * Allows a Commander to inject modifications that must be evaluated at combat time
-   * @param striking - lets the Commander know if it's dealing or taking damage 
-   */
-  public void applyCombatModifiers(BattleParams params, boolean amITheAttacker)
+  /** Called any time you are making a weapon attack */
+  public void buffStrike(StrikeParams params)
+  {}
+  /** Called any time you are attacking a unit, always after {@link #buffStrike(StrikeParams)} */
+  public void buffAttack(BattleParams params)
```

Comment:
Why do we have both a `buffStrike` and a `buffAttack`, and why use these names instead of `applyCombatModifiers`, which feels more descriptive (especially since modifications could be nerfs).
There also needs to be better info here about exactly when each of these is invoked in the combat calculation process so users know when each could/should be used.

Reply:
`buffStrike` exists so that when I slot in attacks vs terrain, we can do conditional buffs that don't require information about exactly what you're shooting.
`buffAttack` exists so you can do conditional buffs that *do* require that info.
`CombatModifier` sounds like a class of object to me.
I thought I *did* put pretty good info in there. It explicitly states strike -> attack -> defense

Reply:
The existing comments specify the order in which they are applied, but not at which phase. They don't give me any idea when I should use one over the other (and less idea of what to use if I want to apply a nerf...).

Reply:
What do you mean by "phase"?
> They don't give me any idea when I should use one over the other

Au contraire: `buffStrike` is for all weapon attacks, `buffAttack` is only for weapon attacks directed at units. `buffDefense` is for when *your* unit is being targeted. That's all there in the comments.
As for nerfing, yeah, it's a little weird nomenclature-wise. I guess I could call them `modifyStrike/Attack/Defense`?

Reply:
For two functions that are named so similarly, more care must be made in delineating the difference, especially since we don't have a handy domain-language dictionary to clarify that "strike" now (as of this PR) just means any targeted weapon discharge, but "attack" means it's directed against a specific *kind* of target (which is currently the only kind of target).
Otherwise it just looks like `buffAttack` is for when attacking a unit, and `buffStrike` is... also for attacking (implicitly: a unit, because what else is there to attack? Even when we get pipe seams/etc, that's not the normative case for weapons employment).

Reply:
> but "attack" means it's directed against a specific kind of target (which is currently the only kind of target).
> Otherwise it just looks like buffAttack is for when attacking a unit, and buffStrike is... also for attacking (implicitly: a unit, because what else is there to attack? Even when we get pipe seams/etc, that's not the normative case for weapons employment).

...and this is kind of why I was wondering whether I should split these changes into two PRs.
I feel like you wouldn't be asking this question if I *had* destructible terrain already implemented here.

I dunno what you want from me. I put the functions right next to each other. I said "use this one to affect attacks on units" and "use that one to affect all attacks".
They have different parameter types (one of which, notably, carried absolutely no data about the thing being targeted until you asked me to re-combine the damage calculation).
They have an explicit call order, too.

Reply:
> I feel like you wouldn't be asking this question if I had destructible terrain already implemented here.

I would. Attacking terrain is definitely something that only even occurs to me in situations where it is needed; it's an exception to the rule of directing bullets at enemy units.
Even if I had taken note of the fine distinction in the two comments (ooh, one includes the word "units" and the other doesn't), that isn't sufficient evidence to determine that `buffStrike` is intended for non-unit-targeted attacks--it's just as likely that you didn't happen to feel like typing "units" at the time you wrote the description. Knowing that both functions are invoked, and in some order, tells me there *should* be a difference, but any inference I'd make based on those javadocs feels like reading tea leaves.

> I dunno what you want from me

Clarity of intent. These function names sound like synonyms; even after being told what they are I only know the difference by rote. At minimum, the javadoc comments should be updated to clarify exactly how their intended use differs--omitted information that might lead to an inference isn't good enough. Just spell it out like the 5-year-old neighbor kid will be the next one using these functions. "This function is intended to be used for any weapon attacks against anything, including destructible map terrain", "This function is intended for adjusting unit firepower when attacking units only, or because the modification depends on the unit being attacked", "This function is called for modifying defense for a unit being attacked, and will be evaluated after `bufStrike` is called for the attacker's Commander, so any firepower adjustment for the guy shooting us is already accounted for in `params` when this function is called" or whatever.

For that matter, is there anything to prevent a Commander from modifying defense in the attack function and vise versa? Maybe we should trim down the functions to take in only the data they need (and maybe also to return the power adjustment instead of modifying `params` directly).

Also also; you say that all three are called, but it's implied that one of them does not apply if we aren't striking a unit. Will `buffAttack` not be called in pipe seam attacks?

Reply:
> Knowing that both functions are invoked, and in some order, tells me there should be a difference, but any inference I'd make based on those javadocs feels like reading tea leaves.

Hmm. Maybe I feel differently because I spend so much time being a pedant.
There are two differences, other than ordering, in those descriptions.
Strike mentions weapons, Attack does not.
Attack mentions hitting a unit, Strike does not.
Seeing as typically, you hit units with weapons, that leaves "hitting a unit" as the most significant difference.
At the same time, Strike *does not rule out* attacking units, ergo it would seem reasonable to assume Strike applies to a strict superset of what Attack applies to.

> omitted information that might lead to an inference isn't good enough

Particularly in writing, I tend to have intent behind the amount and kind of detail I provide. Destructible terrain is one possible use case of `buffStrike`, but I wouldn't want to rule out other potential usecases down the line. I suppose I could just update the doc at that time, but I don't think it makes sense to try to exhaustively account for all use-cases of a given piece of code in the documentation for that piece of code.
The reason the documentation is here instead of there is twofold:
`CombatEngine` itself has a bunch of stuff to keep track of
Someone looking to use these functions will likely care about the detail of general purpose and order of operations.

> For that matter, is there anything to prevent a Commander from modifying defense in the attack function and vise versa?

No. My intent (such as it was) was to use strike/attack/defense as actions units can participate in, rather than the statistics used for that participation *per se*.

> Maybe we should trim down the functions to take in only the data they need

IMO, that's a lot of work for not much gain. If we had 50 people working on the project, I'd be miles more interested in getting the compiler to enforce the separation of concerns/behaviors I'm interested in keeping.
We'd have read-only interfaces for all model objects, and ideally use the compiler to prevent modification of game state anywhere but GameEngine/MapController/GameEvent, and study any that happened in the former two with a critical eye.

> (and maybe also to return the power adjustment instead of modifying params directly).

This kills an existing use case. Venge's Iron Will adjusts his unit's HP value so that he can counterattack as if he hadn't taken damage.
(To tell the truth, I made Venge specifically as a way to effectively stake out and defend the extent of capabilities I wanted our combat calculations to have.)

> Also also; you say that all three are called, but it's implied that one of them does not apply if we aren't striking a unit. Will `buffAttack` not be called in pipe seam attacks?

Actually, only one applies if we aren't striking a unit. `buffStrike` is called unconditionally for all weapon attacks.
IFF the attack targets a unit, `buffAttack` is called... then finally `buffDefense` is called to give the defending unit's CO a look in.
In the case of shooting up the scenery, there's no combatant for `buffAttack` to be interested in, and there's no defending CO to want to call `buffDefense`.

Reply:
> I don't think it makes sense to try to exhaustively account for all use-cases of a given piece of code in the documentation

Agreed. And I agree with documenting the intended use of the functions here, since Commanders are the ones who use them. I don't need exhaustive examples for all use-cases, but I do need enough information to clarify intent for *how* or *when* it should be used. Otherwise I'm left wondering why I have two functions that have synonymous names and take two data types which inherit from each other; it looks like someone added a second version of a function without realizing someone had already implemented it. The comment pointing out the order in which they are called indicates it actually *is* intentional (probably), but doesn't tell me why/how they differ.

> Particularly in writing, I tend to have intent behind the amount and kind of detail I provide.

Nobody who reads your code will read that much into your wording. You probably don't read that much into the wording of other people's code. I don't *want* people trying to read that much into what I write about my code, because as we've (firmly) established, different people think differently, and their natural guess as to what you mean will most likely differ from your intended meaning unless you spell it out.

> Strike mentions weapons, Attack does not.
Attack mentions hitting a unit, Strike does not.
Seeing as typically, you hit units with weapons, that leaves "hitting a unit" as the most significant difference.

Tea leaves. Don't just say one is for hitting units. Clarify that the other is for hitting non-units.

> IMO, that's a lot of work for not much gain

Yeah, probably true.

> Actually, only one applies if we aren't striking a unit. buffStrike is called unconditionally for all weapon attacks.
IFF the attack targets a unit, buffAttack is called... then finally buffDefense is called to give the defending unit's CO a look in.
In the case of shooting up the scenery, there's no combatant for buffAttack to be interested in, and there's no defending CO to want to call buffDefense

Ah, I see. Given that I *still* apparently didn't quite grok this even after all this chatter, it seems like excellent info to include in some documentation...